[spaceship][deliver] add app clips support for specifying default experience metadata

blesson3 opened the pull request:
### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

There is currently no app clips functionality within fastlane. My team and I ran into the issue described in #17639 and had to do manual releases (not fun). I decided to tackle this to help lessen the barrier a bit for others utilizing app clips.

There existed another seemingly abandoned WIP PR #19043 that initially implemented app clips using the web session and private APIs. The ASC API then came out with full app clips support, but with changed models/api-endpoints. This in turn required that PR to re-implement much of it's content, which has not yet been done.

This PR looks to implement the goals of #19043, but with additional `deliver` support so developers can immediately start working with app clips metadata automation via simple `Fastfile` updates.

Resolves #17639

### Description

This PR is one of a few app clips related changes I have made. I am presenting the PRs in smaller pieces so the implications of each PR can be discussed in a tighter scope.

The following app clip default experience functionality has been added to `deliver`:
- option to specify action (PLAY, VIEW, ...)
- option to specify localized subtitle
- option to specify localized default experience header image

If none of these options are specified, then fastlane will copy over the app clip default experience metadata from the live version. This is similar to the default behavior of the ASC UI discussed [here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1130196397).

Example:
```rb
deliver(
  app_clip_default_experience_action: "PLAY",
  app_clip_default_experience_subtitle: {
    "en-US": "this is a test from fastlane",
    "es-MX": "está es una prueba de fastlane"
  },
  app_clip_header_images_path: "./fastlane/app-clip-header-images"
)
```

The localized header images are read from the filesystem same as screenshots:

```sh
$ tree app-clip-header-images/
app-clip-header-images
├── en-US
│   └── random-header-image.jpg
└── es-MX
    └── another-header-image.jpg

2 directories, 2 files
```

These options relate to the ASC UI in the "App Clips" section under a specific App Store version: ![CleanShot 2022-05-19 at 12 12 29@2x](https://user-images.githubusercontent.com/11725484/169361521-73063d31-82ac-44b0-aeab-ae5dcf9b5907.png)

### Discussion

Prior discussion can be [found here](https://github.com/fastlane/fastlane/issues/17639#issuecomment-1121590421).

This PR does NOT implement anything wrt app clip **advanced experiences**. This PR focuses on the default experience associated with the App Store versions only. The advanced experiences are a bit of different beast. If there's interest and it makes sense, I may consider adding that in another PR since I am now familiar with the ASC APIs.

The following should be addressed before this PR is ready to be merged:

1. ~~Apple allows more than one app clip target per app, this PR uses the [first listed app clip only](https://github.com/qatonda/fastlane/blob/9bf2f6bd8bf8dd2bef97dca0e11c6a44928c79a5/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb#L26).~~
  - Resolved in [this comment](https://github.com/fastlane/fastlane/pull/20300#issuecomment-1133162510)

<a href="#discussion-point-2" id="discussion-point-2">#</a> 2.  ~~docs: I see [generated documentation here](https://github.com/fastlane/docs/tree/master/docs/generated), but it's unclear how I update them. I would be happy to update the docs and add an app clips section if someone pointed me in the right direction.~~
  - Resolved with [a commit to update docs](https://github.com/fastlane/fastlane/pull/20300/commits/de5f6d90cc72a7506bc19120cee64a21d4a7c4d4)

3. Do we want the app clips metadata to show up in the interactive HTML metadata report?
    - I have not ever used this feature, but I noticed it while testing.

#### Code Quality

> Disclaimer: I am not a ruby programmer, so I will have missed a lot of ruby-isms in this PR, but I tried my best to keep the code style consistent with the rest of the codebase.

I have not yet added tests for these changes. I have tested them using a small test project and my own projects in production.

Some of the code in this PR has been copied from other relevant parts of the codebase.

##### For example:

The new file [`spaceship/connect_api/models/app_clip_header_image.rb`](https://github.com/fastlane/fastlane/blob/55fb13119ff9b65c865977be6390ecd6cd738360/spaceship/lib/spaceship/connect_api/models/app_clip_header_image.rb) has code copied from [`.../models/app_screenshot.rb`](https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/connect_api/models/app_screenshot.rb) since the upload functionality was the same. I decided not to inherit or subclass because I was sure that making the upload process more general would be more work than it was worth. I am absolutely open to hearing counter arguments.

The PR #17639 has a [similar comment about extracting common functionality here](https://github.com/fastlane/fastlane/pull/19043#pullrequestreview-698633451).

### Testing Steps

#### Basic Testing Setup

1. Update `Gemfile` to 👇 and run `bundle install`, `bundle update fastlane`, or `bundle update`
```rb
gem "fastlane", :git => 'https://github.com/qatonda/fastlane.git', :ref => "matt/app-clips-default-experience"
```
2. Upload a binary to ASC with an app clip target
    - you can follow [a project tutorial provided by Apple here](https://developer.apple.com/documentation/swiftui/fruta_building_a_feature-rich_app_with_swiftui)

#### What to Test

Before specifying any `deliver` options:
- Confirm that fastlane uses the live version's app clip metadata for the current edit version
    - It should copy over the action, subtitle, and header image of the live version
    - This is not already the default when the AS version is created via the ASC APIs, thus the following issue #17639

After specifying the app clip default experience `deliver` options:
- Confirm the app clip default experience metadata specified in the `deliver` options reflects the ASC UI after running `deliver`
- Confirm that fastlane enforces that the app clip header image is exactly 1800x1200 in size
- Confirm that fastlane enforces only one app clip header image per locale
- Confirm that the additional fastlane messages via `UI.{message,important,...}` are appropriate

blesson3 committed:
deliver: add option to specify app clips default experience metadata

blesson3 committed:
deliver: handle app clip metadata copy from live version if no options specified

blesson3 commented on deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:
```diff
@@ -102,5 +120,49 @@ def upload_subtitle_localizations(app_clip_default_experience:, subtitle_localiz
+        # create a default experience and use the live version default experience as a "template"
+        Spaceship::ConnectAPI::AppClipDefaultExperience.create(app_clip_id: app_clip.id, app_store_version_id: edit_version.id, template_default_experience_id: live_version_default_experience.id)
+
+        # TODO: if the above code using the "template" doesn't work, we can use the following to manually copy the values
+        #
+        # default_experience.update(attributes: { action: live_version_default_experience.action })
+        # UI.message("Added app clip default experience action from live version #{live_version.version}")
+
+        # # copy the live version localizations
+        # live_version_localizations = Spaceship::ConnectAPI::AppClipDefaultExperience.get(app_clip_default_experience_id: live_version_default_experience.id, includes: 'appClipDefaultExperienceLocalizations').app_clip_default_experience_localizations
+
+        # live_version_localizations.each do |live_localization|
+        #   attributes = { subtitle: live_localization.subtitle, locale: live_localization.locale }
+        #   Spaceship::ConnectAPI::AppClipDefaultExperienceLocalizations.create(default_experience_id: default_experience.id, attributes: attributes)
+        # end
```
I'm planning to remove this commented code once I can confirm the default experience template option works as I expect.

blesson3 committed:
deliver: fix typo with variables

EmDee commented:
@blesson3 I've ran into an error.

I'm using the latest commit (`028092faaede7d3cc4222618ffa5bcfdb10d4e60`): `Using fastlane 2.206.0 from https://github.com/qatonda/fastlane.git (at matt/app-clips-default-experience@028092f)`

```
bundler: failed to load command: fastlane (/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bin/fastlane)
/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bundler/gems/fastlane-028092faaede/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:96:in `upload_subtitle_localizations': [!] undefined method `id' for nil:NilClass (NoMethodError)
```

blesson3 committed:
deliver: fix issue with creating app clip default experience

fix error in pr comment:
https://github.com/fastlane/fastlane/pull/20300#issuecomment-1132553663

blesson3 commented:
> @blesson3 I've ran into an error.
> 
> I'm using the latest commit (`028092faaede7d3cc4222618ffa5bcfdb10d4e60`): `Using fastlane 2.206.0 from https://github.com/qatonda/fastlane.git (at matt/app-clips-default-experience@028092f)`
> 
> ```
> bundler: failed to load command: fastlane (/Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bin/fastlane)
> /Users/pace/.rbenv/versions/2.6.4/lib/ruby/gems/2.6.0/bundler/gems/fastlane-028092faaede/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:96:in `upload_subtitle_localizations': [!] undefined method `id' for nil:NilClass (NoMethodError)
> ```

@EmDee Thanks for the feedback. I've fixed the issue in 15506706d975e5da7ee095b4af057c2d925907c1. I'm testing the changes now on my own projects to confirm the issue is fixed.

blesson3 committed:
deliver: solidify using the create app clip default experience template option

blesson3 commented:
### Re: apps having more than one app clip target

_(no action needed here on this PR, but this addresses one of the points from the PR description)_

It seems that building an app with more than one app clip target produces the following error from xcode:
```
❌  error: App contains multiple App Clips. It can only have one.
```

It seems impossible to upload a binary with more than one app clip target. What's confusing about this is that the ASC API [allows for more than one app clip to be returned](https://developer.apple.com/documentation/appstoreconnectapi/search_an_app_s_app_clips) for any given app. We handle this by [picking only the first app clip](https://github.com/qatonda/fastlane/blob/9bf2f6bd8bf8dd2bef97dca0e11c6a44928c79a5/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb#L26) from the response. If this changes in the future, we can address it then.

EmDee commented:
> @EmDee Thanks for the feedback. I've fixed the issue in [1550670](https://github.com/fastlane/fastlane/commit/15506706d975e5da7ee095b4af057c2d925907c1) and [fc088f0](https://github.com/fastlane/fastlane/commit/fc088f0e2362fdb2749c8184bd31ad5e946f3c09). I've tested the changes using my own projects and it seems to work well. Please test and lmk!

Thank you, will do! Probably around Tuesday.

EmDee commented:
@blesson3 Looks like you did another rebase?

I was still on the version of 2-3 weeks ago. I just ran into a different error message, not sure, if you've already fixed this one:

```
/Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/spaceship/lib/spaceship/connect_api/tunes/tunes.rb:1291:in `post_app_clip_default_experience': undefined method `data' for #<Hash:0x00007f7d1b86cd20> (NoMethodError)
	from /Users/user/.rbenv/versions/2.6.4/lib/ruby/2.6.0/forwardable.rb:230:in `post_app_clip_default_experience'
	from /Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/spaceship/lib/spaceship/connect_api/models/app_clip_default_experience.rb:25:in `create'
	from /Users/user/.gem/ruby/2.6.0/bundler/gems/fastlane-fc088f0e2362/deliver/lib/deliver/upload_app_clip_default_experience_metadata.rb:51:in `upload_metadata'
```

ConfusedVorlon commented:
This is great. I just hit the relationship error which stopped me submitting a release for an app with an appclip.
I switched to your branch, made no other changes and everything worked great.
(the app used the details from my previous submission)

Brilliant - and Thank You :)

I hope this gets merged soon!

blesson3 committed:
tunes: fix setting body attributes incorrectly

should fix:
https://github.com/fastlane/fastlane/pull/20300#issuecomment-1154433587

EmDee commented:
@blesson3 Great, I'll give the update a test run this Friday.

blesson3 committed:
deliver: remove redundant condition on return

EmDee commented:
@blesson3 Did two releases without a hitch!

EmDee commented:
We've had couple releases now, which all worked great.

@blesson3 Any open tasks on your end or can this be cleaned up and sent to review?

blesson3 commented:
Great to hear. The only thing that might be holding up review is formal documentation. Since app clips is a new class of functionality here, it may be difficult for people to know how to use these new features. Although I'm down to push that into a separate PR so this one can be merged more quickly.

Per [discussion point #2](https://github.com/fastlane/fastlane/pull/20300#discussion-point-2), I was unable to find where I should add documentation. Ultimately I think it should go on [docs.fastlane.tools/actions/deliver](https://docs.fastlane.tools/actions/deliver/).

Pinging maintainers for clarification: @joshdholtz @getaaron

EmDee commented:
Small bump for these great changes and pinging the core team in the hopes that someone can clarify the documentation issue mentioned in https://github.com/fastlane/fastlane/pull/20300#issuecomment-1171764634: @joshdholtz @getaaron @rogerluan @milch

rogerluan commented:
@blesson3 Answering specifically the documentation concerns, the regular parameters/options/descriptions documentation will be autogenerated from the lane metadata. However, I agree with you that extra documentation would be super helpful here, and I'd also expect it to live in the link you mentioned, in which case you'd want to edit `fastlane/lib/fastlane/actions/docs/upload_to_app_store.md.erb` in this repository 😊 it will then be autogenerated in https://github.com/fastlane/docs when creating the next release :)

Hope this helps!

landsman commented:
We are now investigating possibilities how to release App Clips with Fastline and this is probably what we are looking for 🙏

EmDee commented:
Still using this branch without any bigger hitch. 

@blesson3 Would you mind rebasing your branch onto the fastlane's master? Some nice changes have been added, e.g. the use of `altool`.

blesson3 commented:
> @blesson3 Would you mind rebasing your branch onto the fastlane's master? Some nice changes have been added, e.g. the use of altool.

@EmDee Done! Been busy so I haven't yet had the time to invest into writing the documentation.

blesson3 committed:
deliver: add docs for added App Clip metadata options

EmDee commented:
Thanks so much @blesson3!

I can only state again that this MR is pure 🔥

The changes here have saved us so much time in the past couple of months! 🚀

EmDee commented:
@blesson3 Just a quick follow-up question:

We started to add a bunch of languages to our app. As far as I can tell from your changes the `app_clip_default_experience_subtitle` must be set via the `deliver` command, correct? There is no way to have it automatically be populated via some metadata folder structure? (Would definitely be a nice additional follow-up in the future)

blesson3 committed:
deliver: improve App Clip header image docs with a better screenshot

blesson3 commented:
> @blesson3 Just a quick follow-up question:
> 
> We started to add a bunch of languages to our app. As far as I can tell from your changes the `app_clip_default_experience_subtitle` must be set via the `deliver` command, correct? There is no way to have it automatically be populated via some metadata folder structure? (Would definitely be a nice additional follow-up in the future)

@EmDee That's correct in the PR's current implementation. I can definitely explore in the future populating app clip default experience metadata from a folder structure.

blesson3 committed:
deliver: fix bug where app clip action attribute would not update properly

EmDee commented:
@blesson3 Since we are still dependent on these changes, do you mind doing another rebase (as their are some critical fixes from [2.212.1](https://github.com/fastlane/fastlane/releases/tag/2.212.1) we also need.

EmDee commented:
Also maybe we can have this PR looked at, @joshdholtz? Or possibly someone else.

blesson3 commented:
@EmDee Sorry for the delayed response, I've recently moved jobs outside of iOS development and I'm no longer with the company that hosts this PR's source code. I just rebased with `upstream/master` so it should be up to date again.

Feel free to create a new PR using the commits in this one if you want to own this change. I'm not able to do further work or changes based on maintainer feedback (whenever that might be), but I can continue to rebase.

Note that there are several other changes not contained within this PR, like adding an [option to specify beta app clip invocations](https://github.com/qatonda/fastlane/commit/8ac09da3bdc5199b8e7bee0d2976acc86e09f41c) and for [app store review information](https://github.com/qatonda/fastlane/commit/48802b5d5ad56151ea1b72e53f936786ca6bb09a), which can be found in the branch [`qatonda/fastlane:matt/app-clips`](https://github.com/qatonda/fastlane/commits/matt/app-clips). I was planning to open separate PRs after this one was merged. Since I'm not able to see those changes through to master, someone else will have to take over these changes.

EmDee commented:
@blesson3 A great colleague of mine, @telip007, has worked on top of your changes here and fixed all known issues to us. We've opened up a new PR (https://github.com/fastlane/fastlane/pull/21180) where hopefully someone from Fastlane's team will take a look at the changes.

If you want, we can close up this PR here then.
